File: native/android/app/src/main/java/com/tauber/nikonlink/WirelessHttpServer.java
package com.tauber.nikonlink;

import java.io.ByteArrayOutputStream;
import java.io.File;
import java.io.FileOutputStream;
import java.io.InputStream;
import java.io.OutputStream;
import java.net.InetAddress;
import java.net.InetSocketAddress;
import java.net.ServerSocket;
import java.net.Socket;
import java.net.URI;
import java.net.URLDecoder;
import java.nio.charset.StandardCharsets;
import java.util.Base64;
import java.util.Collections;
import java.util.HashMap;
import java.util.Locale;
import java.util.Map;
import java.util.Set;
import java.util.UUID;
import java.util.concurrent.ConcurrentHashMap;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;

final class WirelessHttpServer {
    static final int PORT = 8080;
    private static final long MAXIMUM_FILE_SIZE = 16L * 1024L * 1024L * 1024L;
    private static final int MAXIMUM_HEADER_SIZE = 32 * 1024;

    interface Listener {
        void onStatus(String status);
        void onFileReceived(File file);
        void onError(String message);
    }

    private final File directory;
    private final Listener listener;
    private final ExecutorService clients = Executors.newCachedThreadPool(task -> {
        Thread thread = new Thread(task, "ZENCHE-HTTP-Client");
        thread.setDaemon(true);
        return thread;
    });
    private final Set<Socket> activeClients =
            Collections.newSetFromMap(new ConcurrentHashMap<>());
    private volatile ServerSocket serverSocket;
    private volatile boolean running;

    WirelessHttpServer(File directory, Listener listener) {
        this.directory = directory;
        this.listener = listener;
    }

    synchronized void start() {
        if (running) return;
        running = true;
        Thread thread = new Thread(this::runServer, "ZENCHE-HTTP");
        thread.setDaemon(true);
        thread.start();
    }

    synchronized void stop() {
        running = false;
        ServerSocket server = serverSocket;
        serverSocket = null;
        if (server != null) {
            try {
                server.close();
            } catch (Exception ignored) {
            }
        }
        for (Socket client : activeClients) {
            try {
                client.close();
            } catch (Exception ignored) {
            }
        }
        activeClients.clear();
    }

    private void runServer() {
        ServerSocket server = null;
        try {
            if (!directory.exists() && !directory.mkdirs()) {
                throw new Exception("无法创建无线图片保存目录。");
            }
            server = new ServerSocket();
            serverSocket = server;
            server.setReuseAddress(true);
            server.bind(new InetSocketAddress(InetAddress.getByName("0.0.0.0"), PORT));
            while (running && serverSocket == server) {
                try {
                    Socket client = server.accept();
                    activeClients.add(client);
                    clients.submit(() -> handleClient(client));
                } catch (Exception error) {
                    if (running && serverSocket == server) throw error;
                }
            }
        } catch (Exception error) {
            if (running && serverSocket == server) {
                listener.onError(
                        error.getMessage() == null
                                ? "无法开启 HTTP / WebDAV 图片接收。"
                                : error.getMessage());
            }
        } finally {
            if (serverSocket == server) serverSocket = null;
            if (server != null) {
                try {
                    server.close();
                } catch (Exception ignored) {
                }
            }
        }
    }

    private void handleClient(Socket socket) {
        try (Socket client = socket;
             InputStream input = client.getInputStream();
             OutputStream output = client.getOutputStream()) {
            client.setSoTimeout(120_000);
            Request request = readRequest(input);
            if (request == null) return;

            if (!isAuthenticated(request.headers)) {
                respond(
                        output,
                        401,
                        "Unauthorized",
                        "需要使用 帧澈 ZENCHE 无线收件箱账号。",
                        "WWW-Authenticate: Basic realm=\"ZENCHE\"\r\n");
                return;
            }

            switch (request.method) {
                case "OPTIONS":
                    respond(
                            output,
                            200,
                            "OK",
                            "",
                            "Allow: OPTIONS, GET, PUT, POST, MKCOL, PROPFIND\r\n"
                                    + "DAV: 1\r\n");
                    return;
                case "GET":
                    respond(
                            output,
                            200,
                            "OK",
                            "{\"service\":\"ZENCHE\",\"upload\":\"ready\"}",
                            "Content-Type: application/json; charset=utf-8\r\n");
                    return;
                case "MKCOL":
                    respond(output, 201, "Created", "", "DAV: 1\r\n");
                    return;
                case "PROPFIND":
                    String xml = "<?xml version=\"1.0\" encoding=\"utf-8\"?>"
                            + "<d:multistatus xmlns:d=\"DAV:\">"
                            + "<d:response><d:href>/</d:href><d:propstat><d:prop>"
                            + "<d:resourcetype><d:collection/></d:resourcetype>"
                            + "<d:displayname>ZENCHE</d:displayname>"
                            + "</d:prop><d:status>HTTP/1.1 200 OK</d:status>"
                            + "</d:propstat></d:response></d:multistatus>";
                    respond(
                            output,
                            207,
                            "Multi-Status",
                            xml,
                            "Content-Type: application/xml; charset=utf-8\r\n"
                                    + "DAV: 1\r\n");
                    return;
                case "PUT":
                case "POST":
                    receiveUpload(request, input, output);
                    return;
                default:
                    respond(
                            output,
                            405,
                            "Method Not Allowed",
                            "此无线入口仅支持图片上传。",
                            "Allow: OPTIONS, GET, PUT, POST, MKCOL, PROPFIND\r\n");
            }
        } catch (Exception ignored) {
        } finally {
            activeClients.remove(socket);
        }
    }

    private void receiveUpload(
            Request request,
            InputStream input,
            OutputStream output) throws Exception {
        String lengthText = request.headers.get("content-length");
        if (lengthText == null) {
            respond(output, 411, "Length Required", "请提供 Content-Length。", "");
            return;
        }
        long length;
        try {
            length = Long.parseLong(lengthText);
        } catch (NumberFormatException error) {
            respond(output, 400, "Bad Request", "Content-Length 无效。", "");
            return;
        }
        if (length <= 0 || length > MAXIMUM_FILE_SIZE) {
            respond(output, 413, "Content Too Large", "图片大小必须在 1 字节到 16 GB 之间。", "");
            return;
        }

        String filename = requestedFilename(request);
        if (filename == null || filename.isEmpty()) {
            respond(
                    output,
                    400,
                    "Bad Request",
                    "请使用 /upload/文件名，或提供 X-Filename 请求头。",
                    "");
            return;
        }

        listener.onStatus("正在通过 HTTP / WebDAV 接收 " + filename);
        File destination = uniqueDestination(safeFilename(filename));
        File temporary = new File(directory, "." + UUID.randomUUID() + ".part");
        boolean completed = false;
        try (FileOutputStream file = new FileOutputStream(temporary)) {
            byte[] buffer = new byte[256 * 1024];
            long remaining = length;
            while (remaining > 0) {
                int count = input.read(
                        buffer,
                        0,
                        (int) Math.min(buffer.length, remaining));
                if (count < 0) throw new Exception("图片上传提前中断。");
                if (count == 0) continue;
                file.write(buffer, 0, count);
                remaining -= count;
            }
            file.getFD().sync();
            completed = temporary.renameTo(destination);
            if (!completed) throw new Exception("无法保存无线图片。");
        } finally {
            if (!completed) temporary.delete();
        }

        listener.onFileReceived(destination);
        respond(
                output,
                201,
                "Created",
                "{\"saved\":\"" + jsonEscape(destination.getName()) + "\"}",
                "Content-Type: application/json; charset=utf-8\r\n"
                        + "Location: /" + destination.getName() + "\r\n");
    }

    private Request readRequest(InputStream input) throws Exception {
        ByteArrayOutputStream headerBytes = new ByteArrayOutputStream();
        int matched = 0;
        while (headerBytes.size() < MAXIMUM_HEADER_SIZE) {
            int value = input.read();
            if (value < 0) return null;
            headerBytes.write(value);
            if ((matched == 0 || matched == 2) && value == '\r') {
                matched++;
            } else if ((matched == 1 || matched == 3) && value == '\n') {
                matched++;
                if (matched == 4) break;
            } else {
                matched = value == '\r' ? 1 : 0;
            }
        }
        if (matched != 4) throw new Exception("HTTP 请求头过大。");

        String text = headerBytes.toString(StandardCharsets.ISO_8859_1.name());
        String[] lines = text.substring(0, text.length() - 4).split("\\r\\n");
        if (lines.length == 0) throw new Exception("HTTP 请求无效。");
        String[] first = lines[0].split(" ", 3);
        if (first.length < 2) throw new Exception("HTTP 请求行无效。");
        Map<String, String> headers = new HashMap<>();
        for (int index = 1; index < lines.length; index++) {
            int separator = lines[index].indexOf(':');
            if (separator <= 0) continue;
            headers.put(
                    lines[index].substring(0, separator).trim().toLowerCase(Locale.ROOT),
                    lines[index].substring(separator + 1).trim());
        }
        return new Request(first[0].toUpperCase(Locale.ROOT), first[1], headers);
    }

    private boolean isAuthenticated(Map<String, String> headers) {
        String authorization = headers.get("authorization");
        String expected = "Basic " + Base64.getEncoder().encodeToString(
                (WirelessFtpServer.USERNAME + ":" + WirelessFtpServer.PASSWORD)
                        .getBytes(StandardCharsets.UTF_8));
        return expected.equals(authorization);
    }

    private String requestedFilename(Request request) {
        String explicit = request.headers.get("x-filename");
        if (explicit != null && !explicit.trim().isEmpty()) return explicit.trim();
        try {
            URI uri = URI.create(request.target);
            String path = URLDecoder.decode(
                    uri.getRawPath() == null ? "" : uri.getRawPath(),
                    StandardCharsets.UTF_8.name());
            if (!path.equals("/") && !path.equals("/upload") && !path.equals("/upload/")) {
                return path.substring(path.lastIndexOf('/') + 1);
            }
            String query = uri.getRawQuery();
            if (query != null) {
                for (String part : query.split("&")) {
                    String[] pair = part.split("=", 2);
                    if (pair.length == 2 && pair[0].equalsIgnoreCase("filename")) {
                        return URLDecoder.decode(pair[1], StandardCharsets.UTF_8.name());
                    }
                }
            }
        } catch (Exception ignored) {
        }
        return null;
    }

    private File uniqueDestination(String filename) {
        File initial = new File(directory, filename);
        if (!initial.exists()) return initial;
        int dot = filename.lastIndexOf('.');
        String stem = dot > 0 ? filename.substring(0, dot) : filename;
        String extension = dot > 0 ? filename.substring(dot) : "";
        return new File(directory, stem + "_" + System.currentTimeMillis() + extension);
    }

    private static String safeFilename(String remoteName) {
        String normalized = remoteName.replace('\\', '/').replace('\0', '_');
        String source = normalized.substring(normalized.lastIndexOf('/') + 1).trim();
        String cleaned = source.replaceAll("[<>:\"|?*]", "_");
        return cleaned.isEmpty() || cleaned.equals(".") || cleaned.equals("..")
                ? "NIKON_" + System.currentTimeMillis() + ".JPG"
                : cleaned;
    }

    private static void respond(
            OutputStream output,
            int status,
            String reason,
            String body,
            String extraHeaders) throws Exception {
        byte[] content = body.getBytes(StandardCharsets.UTF_8);
        String headers = "HTTP/1.1 " + status + " " + reason + "\r\n"
                + "Connection: close\r\n"
                + "Content-Length: " + content.length + "\r\n"
                + extraHeaders
                + "\r\n";
        output.write(headers.getBytes(StandardCharsets.ISO_8859_1));
        output.write(content);
        output.flush();
    }

    private static String jsonEscape(String value) {
        return value.replace("\\", "\\\\").replace("\"", "\\\"");
    }

    private static final class Request {
        final String method;
        final String target;
        final Map<String, String> headers;

        Request(String method, String target, Map<String, String> headers) {
            this.method = method;
            this.target = target;
            this.headers = headers;
        }
    }
}

File: native/windows/Services/WirelessHttpServer.cs
using System.IO;
using System.Net;
using System.Net.Sockets;
using System.Text;
using System.Text.Json;

namespace NikonLink.Windows.Services;

public sealed class WirelessHttpServer : IAsyncDisposable
{
    public const int Port = 8080;
    private const long MaximumFileSize = 16L * 1024 * 1024 * 1024;
    private const int MaximumHeaderSize = 32 * 1024;
    private static readonly string Authorization =
        "Basic " + Convert.ToBase64String(
            Encoding.UTF8.GetBytes(
                $"{WirelessFtpServer.Username}:{WirelessFtpServer.Password}"));

    private readonly PhotoLibrary _library;
    private readonly object _clientLock = new();
    private readonly HashSet<TcpClient> _clients = [];
    private TcpListener? _listener;
    private CancellationTokenSource? _cancellation;
    private Task? _acceptTask;

    public WirelessHttpServer(PhotoLibrary library)
    {
        _library = library;
    }

    public event EventHandler<string>? StatusChanged;
    public event EventHandler<string>? FileReceived;
    public event EventHandler<Exception>? Failed;

    public bool IsRunning =>
        _listener is not null &&
        _cancellation is { IsCancellationRequested: false };

    public Task StartAsync()
    {
        if (IsRunning)
        {
            return Task.CompletedTask;
        }
        Directory.CreateDirectory(_library.DirectoryPath);
        _cancellation = new CancellationTokenSource();
        _listener = new TcpListener(IPAddress.Any, Port);
        _listener.Server.SetSocketOption(
            SocketOptionLevel.Socket,
            SocketOptionName.ReuseAddress,
            true);
        _listener.Start();
        _acceptTask = AcceptLoopAsync(_cancellation.Token);
        return Task.CompletedTask;
    }

    public async Task StopAsync()
    {
        var cancellation = _cancellation;
        var listener = _listener;
        _cancellation = null;
        _listener = null;
        cancellation?.Cancel();
        listener?.Stop();

        TcpClient[] clients;
        lock (_clientLock)
        {
            clients = [.. _clients];
            _clients.Clear();
        }
        foreach (var client in clients)
        {
            client.Dispose();
        }
        if (_acceptTask is not null)
        {
            try
            {
                await _acceptTask;
            }
            catch (OperationCanceledException)
            {
            }
            catch (ObjectDisposedException)
            {
            }
        }
        cancellation?.Dispose();
    }

    private async Task AcceptLoopAsync(CancellationToken cancellationToken)
    {
        var listener = _listener;
        if (listener is null)
        {
            return;
        }
        try
        {
            while (!cancellationToken.IsCancellationRequested)
            {
                var client =
                    await listener.AcceptTcpClientAsync(cancellationToken);
                lock (_clientLock)
                {
                    _clients.Add(client);
                }
                _ = HandleClientAsync(client, cancellationToken);
            }
        }
        catch (OperationCanceledException)
        {
        }
        catch (ObjectDisposedException)
        {
        }
        catch (Exception error)
        {
            if (!cancellationToken.IsCancellationRequested)
            {
                Failed?.Invoke(this, error);
            }
        }
    }

    private async Task HandleClientAsync(
        TcpClient client,
        CancellationToken cancellationToken)
    {
        try
        {
            client.ReceiveTimeout = 120_000;
            client.SendTimeout = 120_000;
            client.NoDelay = true;
            await using var stream = client.GetStream();
            var request = await ReadRequestAsync(stream, cancellationToken);
            if (request is null)
            {
                return;
            }
            if (!request.Headers.TryGetValue(
                    "authorization",
                    out var authorization) ||
                authorization != Authorization)
            {
                await RespondAsync(
                    stream,
                    401,
                    "Unauthorized",
                    "需要使用 帧澈 ZENCHE 无线收件箱账号。",
                    new Dictionary<string, string>
                    {
                        ["WWW-Authenticate"] = "Basic realm=\"ZENCHE\""
                    },
                    cancellationToken);
                return;
            }

            switch (request.Method)
            {
                case "OPTIONS":
                    await RespondAsync(
                        stream,
                        200,
                        "OK",
                        string.Empty,
                        new Dictionary<string, string>
                        {
                            ["Allow"] =
                                "OPTIONS, GET, PUT, POST, MKCOL, PROPFIND",
                            ["DAV"] = "1"
                        },
                        cancellationToken);
                    break;
                case "GET":
                    await RespondAsync(
                        stream,
                        200,
                        "OK",
                        "{\"service\":\"ZENCHE\",\"upload\":\"ready\"}",
                        new Dictionary<string, string>
                        {
                            ["Content-Type"] =
                                "application/json; charset=utf-8"
                        },
                        cancellationToken);
                    break;
                case "MKCOL":
                    await RespondAsync(
                        stream,
                        201,
                        "Created",
                        string.Empty,
                        new Dictionary<string, string> { ["DAV"] = "1" },
                        cancellationToken);
                    break;
                case "PROPFIND":
                    const string xml =
                        "<?xml version=\"1.0\" encoding=\"utf-8\"?>" +
                        "<d:multistatus xmlns:d=\"DAV:\"><d:response>" +
                        "<d:href>/</d:href><d:propstat><d:prop>" +
                        "<d:resourcetype><d:collection/></d:resourcetype>" +
                        "<d:displayname>ZENCHE</d:displayname></d:prop>" +
                        "<d:status>HTTP/1.1 200 OK</d:status></d:propstat>" +
                        "</d:response></d:multistatus>";
                    await RespondAsync(
                        stream,
                        207,
                        "Multi-Status",
                        xml,
                        new Dictionary<string, string>
                        {
                            ["Content-Type"] =
                                "application/xml; charset=utf-8",
                            ["DAV"] = "1"
                        },
                        cancellationToken);
                    break;
                case "PUT":
                case "POST":
                    await ReceiveAsync(request, stream, cancellationToken);
                    break;
                default:
                    await RespondAsync(
                        stream,
                        405,
                        "Method Not Allowed",
                        "此无线入口仅支持图片上传。",
                        new Dictionary<string, string>
                        {
                            ["Allow"] =
                                "OPTIONS, GET, PUT, POST, MKCOL, PROPFIND"
                        },
                        cancellationToken);
                    break;
            }
        }
        catch (OperationCanceledException)
        {
        }
        catch (IOException)
        {
        }
        catch (Exception error)
        {
            if (!cancellationToken.IsCancellationRequested)
            {
                Failed?.Invoke(this, error);
            }
        }
        finally
        {
            client.Dispose();
            lock (_clientLock)
            {
                _clients.Remove(client);
            }
        }
    }

    private async Task ReceiveAsync(
        Request request,
        NetworkStream stream,
        CancellationToken cancellationToken)
    {
        if (!request.Headers.TryGetValue(
                "content-length",
                out var lengthText) ||
            !long.TryParse(lengthText, out var length))
        {
            await RespondAsync(
                stream,
                411,
                "Length Required",
                "请提供有效的 Content-Length。",
                null,
                cancellationToken);
            return;
        }
        if (length <= 0 || length > MaximumFileSize)
        {
            await RespondAsync(
                stream,
                413,
                "Content Too Large",
                "图片大小必须在 1 字节到 16 GB 之间。",
                null,
                cancellationToken);
            return;
        }
        var filename = RequestedFilename(request);
        if (string.IsNullOrWhiteSpace(filename))
        {
            await RespondAsync(
                stream,
                400,
                "Bad Request",
                "请使用 /upload/文件名，或提供 X-Filename 请求头。",
                null,
                cancellationToken);
            return;
        }

        StatusChanged?.Invoke(
            this,
            $"正在通过 HTTP / WebDAV 接收 {filename}");
        var destination = _library.UniqueDestination(filename);
        var temporary =
            destination + "." + Guid.NewGuid().ToString("N") + ".part";
        try
        {
            await using var output = new FileStream(
                temporary,
                FileMode.CreateNew,
                FileAccess.Write,
                FileShare.None,
                256 * 1024,
                FileOptions.Asynchronous | FileOptions.SequentialScan);
            var buffer = new byte[256 * 1024];
            long remaining = length;
            while (remaining > 0)
            {
                var count = await stream.ReadAsync(
                    buffer.AsMemory(
                        0,
                        (int)Math.Min(buffer.Length, remaining)),
                    cancellationToken);
                if (count == 0)
                {
                    throw new IOException("图片上传提前中断。");
                }
                await output.WriteAsync(
                    buffer.AsMemory(0, count),
                    cancellationToken);
                remaining -= count;
            }
            await output.FlushAsync(cancellationToken);
            output.Flush(true);
            File.Move(temporary, destination);
        }
        finally
        {
            if (File.Exists(temporary))
            {
                File.Delete(temporary);
            }
        }

        FileReceived?.Invoke(this, destination);
        var body = JsonSerializer.Serialize(
            new { saved = Path.GetFileName(destination) });
        await RespondAsync(
            stream,
            201,
            "Created",
            body,
            new Dictionary<string, string>
            {
                ["Content-Type"] = "application/json; charset=utf-8",
                ["Location"] =
                    "/" + Uri.EscapeDataString(Path.GetFileName(destination))
            },
            cancellationToken);
    }

    private static async Task<Request?> ReadRequestAsync(
        NetworkStream stream,
        CancellationToken cancellationToken)
    {
        var bytes = new List<byte>();
        var single = new byte[1];
        var matched = 0;
        while (bytes.Count < MaximumHeaderSize)
        {
            var count = await stream.ReadAsync(single, cancellationToken);
            if (count == 0)
            {
                return null;
            }
            var value = single[0];
            bytes.Add(value);
            if ((matched is 0 or 2) && value == '\r')
            {
                matched++;
            }
            else if ((matched is 1 or 3) && value == '\n')
            {
                matched++;
                if (matched == 4)
                {
                    break;
                }
            }
            else
            {
                matched = value == '\r' ? 1 : 0;
            }
        }
        if (matched != 4)
        {
            return null;
        }
        var text = Encoding.UTF8.GetString(
            bytes.GetRange(0, bytes.Count - 4).ToArray());
        var lines = text.Split("\r\n", StringSplitOptions.None);
        var requestLine = lines[0].Split(' ', 3);
        if (requestLine.Length < 2)
        {
            return null;
        }
        var headers = new Dictionary<string, string>(
            StringComparer.OrdinalIgnoreCase);
        foreach (var line in lines.Skip(1))
        {
            var separator = line.IndexOf(':');
            if (separator <= 0)
            {
                continue;
            }
            headers[line[..separator].Trim().ToLowerInvariant()] =
                line[(separator + 1)..].Trim();
        }
        return new Request(
            requestLine[0].ToUpperInvariant(),
            requestLine[1],
            headers);
    }

    private static string? RequestedFilename(Request request)
    {
        if (request.Headers.TryGetValue("x-filename", out var explicitName) &&
            !string.IsNullOrWhiteSpace(explicitName))
        {
            return explicitName.Trim();
        }
        if (!Uri.TryCreate(
                "http://127.0.0.1" + request.Target,
                UriKind.Absolute,
                out var uri))
        {
            return null;
        }
        var path = Uri.UnescapeDataString(uri.AbsolutePath);
        if (path is not "/" and not "/upload" and not "/upload/")
        {
            return Path.GetFileName(path);
        }
        foreach (var part in uri.Query.TrimStart('?').Split('&'))
        {
            var pair = part.Split('=', 2);
            if (pair.Length == 2 &&
                pair[0].Equals("filename", StringComparison.OrdinalIgnoreCase))
            {
                return WebUtility.UrlDecode(pair[1]);
            }
        }
        return null;
    }

    private static async Task RespondAsync(
        NetworkStream stream,
        int status,
        string reason,
        string body,
        IReadOnlyDictionary<string, string>? headers,
        CancellationToken cancellationToken)
    {
        var content = Encoding.UTF8.GetBytes(body);
        var head = new StringBuilder()
            .Append($"HTTP/1.1 {status} {reason}\r\n")
            .Append("Connection: close\r\n")
            .Append($"Content-Length: {content.Length}\r\n");
        if (headers is not null)
        {
            foreach (var (name, value) in headers)
            {
                head.Append($"{name}: {value}\r\n");
            }
        }
        head.Append("\r\n");
        await stream.WriteAsync(
            Encoding.ASCII.GetBytes(head.ToString()),
            cancellationToken);
        await stream.WriteAsync(content, cancellationToken);
        await stream.FlushAsync(cancellationToken);
    }

    private sealed record Request(
        string Method,
        string Target,
        Dictionary<string, string> Headers);

    public async ValueTask DisposeAsync()
    {
        await StopAsync();
        GC.SuppressFinalize(this);
    }
}
